Multi edge trigger

ggggggggg opened the pull request:
This adds an experimental `EdgeMulti` triggering mode with tests. It's incompatible with other trigger modes, and only triggers on rising edges, but I think it's worth merging in so we can test it with actual data sources. It's works basically how we discussed
1. Look for a difference between two points larger than a threshold. Call this i_potential. In my testing offline I found that if you do the kink model, it doesn't matter much how you got i_potential, so I used the easiest method.
2. Look for a local maximum following i_potential.
3. Use  6 samples before and 3 samples after i_potential to fit a kink model looking at values -1,-0.5-,0.5,1 relative to i_potential. Note kbest as a trigger.
4. Go through all sets of t,u,v where t is the index of the last trigger, u is the current trigger, v is the next trigger, and make records. There are a couple of settings, but the MakeShortRecords setting makes records greedy after a trigger, and makes records as large as possible with no shared samples.

It would be pretty easy to support negative going edges, and it would fairly easy to support auto triggers of some sort coexisting.

`AnalyzeData` works with records of `length<=dsp.NSamples`. Note that the standard analysis quantities are almost all NaN when `rec.presamples=0`

joefowler commented on process_data.go:
```diff
@@ -206,22 +185,29 @@ func (dsp *DataStreamProcessor) AnalyzeData(records []*DataRecord) {
+			k := nbases
+			j := truncFront
+			l := cols - truncBack
+			projectors := dsp.projectors.Slice(i, k, j, l)
```
Problem: the truncation of a basis does not produce projectors that are truncations of the originals. Our choices are:

1. Don't (yet) allow variable-length records. I know they will be important someday, but I am afraid that learning/deciding how to handle them might bog down this project considerably.

1. Have Dastard store projectors for all allowed lengths. (This might mean that we designate a set of allowed lengths, so that 500, 520, 540 are all allowed, while a record that you might want to be of length 534 has to be truncated to 520.)

1. Perform whitening or approximate whitening on the data in dastard. If you do this, then I think projection is still not as simple as truncating the projector, but it might be somewhat simpler.

ggggggggg replied:
I thought you told me we could just truncate, but given that we can't, you're right so I took this out. I left in all the changes that make the summary quantities work with variable length records, and the tests on those quantities.

joefowler commented on triggering_test.go:
```diff
@@ -456,3 +583,21 @@ func BenchmarkLevelTrigger0TriggersOpsAreSamples(b *testing.B) {
+	}
+	ymodel, a, b, c, X2, err = kinkModelResult(4, xdata, ydata)
+	if a != 0.6818181818181821 || b != 0.22727272727272738 ||
+		c != 1.1363636363636362 || X2 != 0.45454545454545453 || err != nil {
```
It seems sketchy to do tests on floating point equality. I don't see any go-supported test for approximate equality, though, not in the standard libraries. Not sure what to suggest.

ggggggggg replied:
I agree in principle, but this work, and is simple, and they don't provide anything to help. I guess we could write our own approx tester or find one in a package, but lets wait till its a problem.

joefowler commented on triggering.go:
```diff
@@ -4,24 +4,269 @@ import (
+	}
+}
+
+// takes k, an index into xdata and into ydata
```
Need a clearer comment, especially as to the meaning of k. I assume it's the real-number location of the kink, but I don't think the above line says that.

Also does't golint insist that comments be of the form?
```go
// kinkModelResult does x,y,z...
// additional lines...
```
Actually, that might only be comments on exported functions.

ggggggggg commented:
I pushed a commit that addressed your comments. Thanks for the feedback. I meant to split my commit into a few with different message, but I messed that part up.
